ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/RaoInputHelper.java

```diff
@@ -83,44 +85,76 @@ public static void synchronize(Crac crac, Network network) {
                 }
+                // do not delete contingencies now as they are needed to check which associated states/cnecs should be removed as well
             });
         }
 
-        absentFromNetworkContingencies.forEach(contingency ->  {
-            crac.getStatesFromContingency(contingency.getId()).forEach(state -> {
+        // list States whose contingency does not exist anymore
+        Set<State> removedStates = new HashSet<>();
+        for (Contingency contingency : removedContingencies) {
+            crac.getStates(contingency).forEach(state -> {
+                report.add(String.format("[REMOVED] State %s is removed because its associated contingency [%s] has been removed", state.getId(), contingency.getId()));
+                removedStates.add(state);
+            });
+            // do not delete states now as they are needed to check which associated ra/cnecs should be removed as well
+        }
+
+        // remove Cnec whose contingency does not exist anymore
+        removedContingencies.forEach(contingency ->
+            crac.getStatesFromContingency(contingency.getId()).forEach(state ->
                 crac.getCnecs(state).forEach(cnec -> {
                     crac.removeCnec(cnec.getId());
                     report.add(String.format("[REMOVED] Cnec %s is removed because its associated contingency [%s] has been removed", cnec.getId(), contingency.getId()));
-                });
-                crac.removeState(state.getId());
-            });
-            crac.removeContingency(contingency.getId());
-        });
+                })
+            )
+        );
 
         // remove Remedial Action with an empty list of NetworkElement
-        ArrayList<NetworkAction> noValidAction = new ArrayList<>();
+        Set<RemedialAction<?>> noValidAction = new HashSet<>();
         crac.getNetworkActions().stream().filter(na -> na.getNetworkElements().isEmpty()).forEach(na -> {
             report.add(String.format("[REMOVED] Remedial Action %s has no associated action. It is removed from the Crac", na.getId()));
             noValidAction.add(na);
         });
+        crac.getRangeActions().stream().filter(ra -> ra.getNetworkElements().isEmpty()).forEach(ra -> {
+            report.add(String.format("[REMOVED] Remedial Action %s has no associated action. It is removed from the Crac", ra.getId()));
+            noValidAction.add(ra);
+        });
         noValidAction.forEach(networkAction -> crac.removeNetworkAction(networkAction.getId()));
 
-        // todo : remove State whose contingency does not exist anymore
-        // todo : remove onState remedial action whose state does not exist anymore
+        // remove On State usage rule with invalid state
+        crac.getRangeActions().forEach(ra -> checkUsageRules(ra, removedStates, report));
+        crac.getNetworkActions().forEach(ra -> checkUsageRules(ra, removedStates, report));
```

Comment:
should we remove remedial actions with no usage rules left after this ? (if the RA only had one OnState usage rule that was removed by checkUsageRules)

Reply:
This is something i was considering, but Farao actually handle well the case where a RemedialAction has no usage Rule : it will consider it UNAVAILABLE for all states.

I therefore prefer to keep these remedial actions in the Crac, then we still have it in the results and we can see that it was never activated. I allows to keep some more consistency between input/output without risking a crash of the RAO

Reply:
okay !